File: netlify/functions/optimize.py
import json
import numpy as np
from scipy.optimize import minimize

def handler(event, context):
    """
    Netlify Function to optimize soil mixing ratios.

    Expected input (JSON):
    {
        "batches": [
            {"name": "Batch 1", "pH": 7.2, "Arsenic": 11, ...},
            {"name": "Batch 2", "pH": 9, "Arsenic": 22, ...},
            ...
        ],
        "limits": {
            "pH": {"lower": 5.5, "upper": 8.5},
            "Arsenic": {"lower": 0, "upper": 37},
            ...
        },
        "tolerance": 0.75  # optional, default 0.75
    }

    Returns:
    {
        "success": true/false,
        "ratios": [0.35, 0.45, 0.20],
        "blended_values": {"pH": 7.5, "Arsenic": 15, ...},
        "residuals": {"pH": 0.12, "Arsenic": 0.23, ...},
        "total_residual": 0.45,
        "within_tolerance": true/false,
        "suggested_tolerance": 0.85  # if no solution found
    }
    """

    # Handle CORS preflight
    headers = {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS',
        'Content-Type': 'application/json'
    }

    if event.get('httpMethod') == 'OPTIONS':
        return {
            'statusCode': 200,
            'headers': headers,
            'body': ''
        }

    try:
        # Parse request body
        body = json.loads(event.get('body', '{}'))
        batches = body.get('batches', [])
        limits = body.get('limits', {})
        tolerance = body.get('tolerance', 0.75)

        if not batches or not limits:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'Missing batches or limits data'})
            }

        # Run optimization
        result = optimize_mix(batches, limits, tolerance)

        return {
            'statusCode': 200,
            'headers': headers,
            'body': json.dumps(result)
        }

    except Exception as e:
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': str(e)})
        }


def optimize_mix(batches, limits, tolerance=0.75):
    """
    Optimize soil mixing ratios using scipy.optimize.

    Args:
        batches: List of dicts with parameter values
        limits: Dict with upper/lower for each parameter
        tolerance: Target tolerance (0.0-1.0) from midpoint

    Returns:
        Dict with optimization results
    """
    n_batches = len(batches)

    # Get all parameter names (excluding 'name' field)
    param_names = [key for key in batches[0].keys() if key != 'name']

    # Filter out parameters with upper limit = 9999 (ignore)
    active_params = [p for p in param_names if limits.get(p, {}).get('upper', 9999) != 9999]

    # Calculate midpoints and ranges for active parameters
    midpoints = {}
    ranges = {}
    for param in active_params:
        lower = limits[param]['lower']
        upper = limits[param]['upper']
        midpoints[param] = (upper + lower) / 2
        ranges[param] = upper - lower

        # Avoid division by zero
        if ranges[param] == 0:
            ranges[param] = 1e-10

    # Objective function: minimize sum of normalized residuals
    def objective(ratios):
        total_residual = 0
        for param in active_params:
            blended = sum(ratios[i] * batches[i][param] for i in range(n_batches))
            residual = abs(blended - midpoints[param]) / ranges[param]
            total_residual += residual
        return total_residual

    # Constraints
    constraints = [
        {'type': 'eq', 'fun': lambda r: sum(r) - 1.0}  # ratios sum to 1
    ]

    # Add hard limit constraints for each parameter
    for param in active_params:
        upper = limits[param]['upper']
        lower = limits[param]['lower']

        # Upper limit constraint
        constraints.append({
            'type': 'ineq',
            'fun': lambda r, p=param, u=upper: u - sum(r[i] * batches[i][p] for i in range(n_batches))
        })

        # Lower limit constraint
        constraints.append({
            'type': 'ineq',
            'fun': lambda r, p=param, l=lower: sum(r[i] * batches[i][p] for i in range(n_batches)) - l
        })

    # Bounds: each ratio between 0 and 1
    bounds = [(0, 1) for _ in range(n_batches)]

    # Initial guess: equal proportions
    x0 = np.array([1/n_batches] * n_batches)

    # Run optimization
    opt_result = minimize(
        objective,
        x0,
        method='SLSQP',
        bounds=bounds,
        constraints=constraints,
        options={'maxiter': 1000, 'ftol': 1e-9}
    )

    # Calculate blended values and residuals with optimal ratios
    optimal_ratios = opt_result.x
    blended_values = {}
    residuals = {}
    within_limits = True
    within_tolerance_count = 0

    for param in param_names:
        blended = sum(optimal_ratios[i] * batches[i][param] for i in range(n_batches))
        blended_values[param] = blended

        if param in active_params:
            lower = limits[param]['lower']
            upper = limits[param]['upper']
            midpoint = midpoints[param]
            param_range = ranges[param]

            # Calculate normalized residual
            residual = abs(blended - midpoint) / param_range
            residuals[param] = residual

            # Check if within limits
            if blended < lower or blended > upper:
                within_limits = False

            # Check if within tolerance
            tolerance_range = param_range * (1 - tolerance) / 2
            if abs(blended - midpoint) <= tolerance_range:
                within_tolerance_count += 1

    total_residual = sum(residuals.values())
    within_tolerance = within_tolerance_count == len(active_params)

    # If no solution within tolerance, suggest relaxed tolerance
    suggested_tolerance = None
    if not within_tolerance and opt_result.success:
        # Calculate what tolerance would be needed
        max_residual = max(residuals.values()) if residuals else 0
        # tolerance_range / param_range = (1 - tolerance) / 2
        # max_residual = tolerance_range / param_range
        # max_residual = (1 - tolerance) / 2
        # tolerance = 1 - 2 * max_residual
        suggested_tolerance = max(0.0, min(1.0, 1 - 2 * max_residual))
        suggested_tolerance = round(suggested_tolerance, 2)

    return {
        'success': opt_result.success and within_limits,
        'ratios': [round(r, 4) for r in optimal_ratios.tolist()],
        'blended_values': {k: round(v, 4) for k, v in blended_values.items()},
        'residuals': {k: round(v, 4) for k, v in residuals.items()},
        'total_residual': round(total_residual, 4),
        'within_tolerance': within_tolerance,
        'within_limits': within_limits,
        'suggested_tolerance': suggested_tolerance,
        'message': opt_result.message if hasattr(opt_result, 'message') else ''
    }

File: netlify/functions/optimize.js
/**
 * Netlify Function to optimize soil mixing ratios using JavaScript
 * Implements the same algorithm as the Python version but in Node.js
 */

export const handler = async (event, context) => {
  // Handle CORS preflight
  const headers = {
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
    'Access-Control-Allow-Methods': 'POST, OPTIONS',
    'Content-Type': 'application/json'
  };

  if (event.httpMethod === 'OPTIONS') {
    return {
      statusCode: 200,
      headers,
      body: ''
    };
  }

  try {
    // Parse request body
    const body = JSON.parse(event.body || '{}');
    const batches = body.batches || [];
    const limits = body.limits || {};
    const tolerance = body.tolerance || 0.75;

    if (!batches || batches.length === 0 || !limits || Object.keys(limits).length === 0) {
      return {
        statusCode: 400,
        headers,
        body: JSON.stringify({ error: 'Missing batches or limits data' })
      };
    }

    // Run optimization
    const result = optimizeMix(batches, limits, tolerance);

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify(result)
    };

  } catch (error) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({
        error: error.message,
        stack: error.stack
      })
    };
  }
}

/**
 * Optimize soil mixing ratios using gradient descent with Sequential Least Squares Programming (SLSQP) approach
 */
function optimizeMix(batches, limits, tolerance = 0.75) {
  const nBatches = batches.length;

  // Get all parameter names (excluding 'name' field)
  const paramNames = Object.keys(batches[0]).filter(key => key !== 'name');

  // Identify parameters with missing data (null values in any batch)
  const missingDataParams = paramNames.filter(param =>
    batches.some(batch => batch[param] === null || batch[param] === undefined)
  );

  // Filter out parameters with upper limit = 9999 (ignore) AND parameters with missing data
  const activeParams = paramNames.filter(p =>
    limits[p]?.upper !== 9999 && !missingDataParams.includes(p)
  );

  // Calculate targets and ranges for active parameters
  const targets = {};
  const ranges = {};
  activeParams.forEach(param => {
    const lower = limits[param].lower;
    const upper = limits[param].upper;

    // Zero-seeking: if lower limit is 0, aim for 0 (minimize contaminants)
    // Otherwise, aim for midpoint (balance between limits)
    if (lower === 0) {
      targets[param] = 0;  // Minimize contaminants to zero
    } else {
      targets[param] = (upper + lower) / 2;  // Target midpoint for non-contaminants
    }

    ranges[param] = upper - lower || 1e-10;  // Avoid division by zero
  });

  // Objective function: minimize sum of normalized residuals from TARGET
  function objective(ratios) {
    let totalResidual = 0;
    activeParams.forEach(param => {
      const blended = ratios.reduce((sum, ratio, i) => sum + ratio * batches[i][param], 0);
      const residual = Math.abs(blended - targets[param]) / ranges[param];
      totalResidual += residual;
    });
    return totalResidual;
  }

  // Check if ratios satisfy all constraints
  function checkConstraints(ratios) {
    // Check sum to 1
    const sum = ratios.reduce((a, b) => a + b, 0);
    if (Math.abs(sum - 1.0) > 0.001) return false;

    // Check all ratios >= 0 and <= 1
    if (ratios.some(r => r < -1e-6 || r > 1 + 1e-6)) return false;

    // Check all parameters within limits
    for (const param of activeParams) {
      const blended = ratios.reduce((sum, ratio, i) => sum + ratio * batches[i][param], 0);
      const lower = limits[param].lower;
      const upper = limits[param].upper;
      if (blended < lower - 1e-6 || blended > upper + 1e-6) {
        return false;
      }
    }

    return true;
  }

  // Gradient descent with projection (matching client-side algorithm)
  function optimize() {
    // Start with equal proportions
    let ratios = new Array(nBatches).fill(1 / nBatches);
    const learningRate = 0.01;
    const maxIterations = 5000;
    let bestRatios = [...ratios];
    let bestObjective = objective(ratios);

    for (let iter = 0; iter < maxIterations; iter++) {
      // Calculate gradient numerically
      const gradient = new Array(nBatches).fill(0);
      const epsilon = 0.0001;

      for (let i = 0; i < nBatches; i++) {
        const ratiosPlus = [...ratios];
        ratiosPlus[i] += epsilon;
        const objPlus = objective(ratiosPlus);
        gradient[i] = (objPlus - objective(ratios)) / epsilon;
      }

      // Update ratios using gradient
      const newRatios = ratios.map((r, i) => r - learningRate * gradient[i]);

      // Project onto constraints (sum = 1, all >= 0)
      const boundedRatios = newRatios.map(r => Math.max(0, r));
      const sum = boundedRatios.reduce((a, b) => a + b, 0);
      const projectedRatios = boundedRatios.map(r => r / (sum || 1));

      // Check if this is better AND satisfies constraints
      const currentObj = objective(projectedRatios);
      if (currentObj < bestObjective && checkConstraints(projectedRatios)) {
        bestObjective = currentObj;
        bestRatios = [...projectedRatios];
      }

      // ALWAYS update ratios (don't get stuck!)
      ratios = projectedRatios;

      // Early stopping if objective is very small
      if (currentObj < 0.001) break;
    }

    // If gradient descent didn't find valid solution, try exhaustive search
    if (!checkConstraints(bestRatios) && nBatches <= 3) {
      const searchResult = exhaustiveSearch(nBatches);
      if (searchResult && checkConstraints(searchResult)) {
        bestRatios = searchResult;
        bestObjective = objective(searchResult);
      }
    }

    return bestRatios;
  }

  // Exhaustive search for small number of batches
  function exhaustiveSearch(n) {
    let best = null;
    let bestObj = Infinity;

    // Try combinations in steps of 0.05
    const step = 0.05;
    function search(ratios, remaining, depth) {
      if (depth === n - 1) {
        ratios[depth] = remaining;
        if (Math.abs(remaining - Math.round(remaining / step) * step) < 0.01) {
          if (checkConstraints(ratios)) {
            const obj = objective(ratios);
            if (obj < bestObj) {
              bestObj = obj;
              best = [...ratios];
            }
          }
        }
        return;
      }

      for (let r = 0; r <= remaining; r += step) {
        ratios[depth] = r;
        search(ratios, remaining - r, depth + 1);
      }
    }

    search(new Array(n).fill(0), 1.0, 0);
    return best;
  }

  // Run optimization
  const optimalRatios = optimize();

  // Calculate blended values and residuals with optimal ratios
  const blendedValues = {};
  const residuals = {};
  let withinLimits = true;
  let withinToleranceCount = 0;

  paramNames.forEach(param => {
    const blended = optimalRatios.reduce((sum, ratio, i) => {
      const value = batches[i][param];
      return sum + ratio * (value !== null && value !== undefined ? value : 0);
    }, 0);
    blendedValues[param] = blended;

    if (activeParams.includes(param)) {
      const lower = limits[param].lower;
      const upper = limits[param].upper;
      const target = targets[param];
      const paramRange = ranges[param];

      // Calculate normalized residual from TARGET
      const residual = Math.abs(blended - target) / paramRange;
      residuals[param] = residual;

      // Check if within limits
      if (blended < lower || blended > upper) {
        withinLimits = false;
      }

      // Check if within tolerance
      const toleranceRange = paramRange * (1 - tolerance) / 2;
      if (Math.abs(blended - target) <= toleranceRange) {
        withinToleranceCount++;
      }
    }
  });

  const totalResidual = Object.values(residuals).reduce((a, b) => a + b, 0);
  const withinTolerance = withinToleranceCount === activeParams.length;

  // If no solution within tolerance, suggest relaxed tolerance
  let suggestedTolerance = null;
  if (!withinTolerance && withinLimits) {
    const maxResidual = Math.max(...Object.values(residuals));
    suggestedTolerance = Math.max(0.0, Math.min(1.0, 1 - 2 * maxResidual));
    suggestedTolerance = Math.round(suggestedTolerance * 100) / 100;
  }

  return {
    success: withinLimits,
    ratios: optimalRatios.map(r => Math.round(r * 10000) / 10000),
    blended_values: Object.fromEntries(
      Object.entries(blendedValues).map(([k, v]) => [k, Math.round(v * 10000) / 10000])
    ),
    residuals: Object.fromEntries(
      Object.entries(residuals).map(([k, v]) => [k, Math.round(v * 10000) / 10000])
    ),
    total_residual: Math.round(totalResidual * 10000) / 10000,
    within_tolerance: withinTolerance,
    within_limits: withinLimits,
    suggested_tolerance: suggestedTolerance,
    missing_data_params: missingDataParams,
    message: withinLimits ? 'Optimisation successful' : 'No valid solution found within limits'
  };
}
